### PR TITLE
Refactoring and request future cancellation

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
- Refactoring - moving the request execution logic to a separate class 
- Invoke `responseHandler#onError` when the request future is cancelled

There are two commits, easier to review one by one

## Testing
Added test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: http-clients/aws-crt-client/src/test/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutorTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import software.amazon.awssdk.crt.CrtRuntimeException;
+import software.amazon.awssdk.crt.http.HttpClientConnection;
+import software.amazon.awssdk.crt.http.HttpClientConnectionManager;
+import software.amazon.awssdk.crt.http.HttpRequest;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+
+@RunWith(MockitoJUnitRunner.class)
+public class CrtRequestExecutorTest {
```
Comment (Quanzzzz) [inline http-clients/aws-crt-client/src/test/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutorTest.java:43]: Do we need a WireMock test here to test the situation when the request is executed successfully?

File: test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingNetworkTrafficListener.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import com.github.tomakehurst.wiremock.http.trafficlistener.WiremockNetworkTrafficListener;
+import java.net.Socket;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Simple implementation of {@link WiremockNetworkTrafficListener} to record all requests received as a string for later
+ * verification.
+ */
+public class RecordingNetworkTrafficListener implements WiremockNetworkTrafficListener {
```
Comment (Quanzzzz) [inline test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingNetworkTrafficListener.java:27]: Question: Where do we use these two classes: `RecordingNetworkTrafficListener` and `RecordingResponseHandler`? Are they for the normal case test?

File: test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingResponseHandler.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.http.async.SdkAsyncHttpResponseHandler;
+import software.amazon.awssdk.http.async.SimpleSubscriber;
+import software.amazon.awssdk.metrics.MetricCollector;
+
+public final class RecordingResponseHandler implements SdkAsyncHttpResponseHandler {
```
Comment (Quanzzzz) [inline test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingResponseHandler.java:28]: Same question as above.

File: http-clients/aws-crt-client/src/test/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutorTest.java (hunk 1 above)
Comment (zoewangg) [reply http-clients/aws-crt-client/src/test/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutorTest.java:43]: This test is not using wiremock, but yeah, I can add a test for happy path

File: test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingNetworkTrafficListener.java (hunk 2 above)
Comment (zoewangg) [reply test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingNetworkTrafficListener.java:27]: Yeah, they are used in the wiremock tests.

File: test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingResponseHandler.java (hunk 3 above)
Comment (zoewangg) [reply test/http-client-tests/src/main/java/software/amazon/awssdk/http/RecordingResponseHandler.java:28]: Yeah, they are only used in tests.

**Review by Quanzzzz — APPROVED** [review]

**Pull request merged**